### PR TITLE
chromium: Add optional vaapi hw video decoding support

### PR DESCRIPTION
**Pull request by amazari**
Add 'VAAPISupport' parameter to chromium expression. If set to true, a
patch enabling hw video decoding through vaapi is applied.

The patch originates from Arch Linux
https://git.archlinux.org/svntogit/packages.git/plain/trunk/chromium-vaapi.patch?h=packages/chromium

Tested on 72.xx

###### Motivation for this change
Using HW video decoding should improve performances and "smoothness" while reducing battery consumption.
Seems to be a requested feature per issue #21481

###### Things done
Import a patch from Arch and enable its application behind the boolean VAAPISupport parameter for the chromium function (default = false).
Naming of the parameter is debatable as there seems to be 2 conventions: xxxSupport and enableXxx.
Also should the value be conditioned by the presence of the libva package and/or the presence of libva in opengl libs?
Should an assertion be added to ensure that the version of chromium is equal to 72.xx ?

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by hyperfekt** [comment]
Are you sure this is working? (Easiest way is probably to observe a YouTube 4K video which is served as VP9 and will be *very* noticeable if done on the CPU.)
There are a few patches missing that Arch Linux uses that are only in Chromium upstream with version 73: https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/chromium&id=53de8f9a9b72a1cf7b0438a109240104d1303eb6#n96
Since we missed the cutoff date for 19.03 (I had higher priorities unfortunately), I suggest we wait until 73 is released (estimated for the 12th) so it's cleaner.
I would also suggest following the example of all the other distros and enabling the patch by default if it works, since we really don't want people to have to build their own version of Chromium.

**Comment by colemickens** [comment]
As a small note, in 73 and up, this is as simple as enabling `use_vaapi=true` as a GN flag: https://github.com/chromium/chromium/commit/31225b9c5f3f685d65f742dc129241c30c32157c

(in the GH UI, you can see that that commit is present in 73+ builds).

Also, if I had to guess, as a result, the strategy in this PR might not work for beta/dev flavors of the chromium package.

**Comment by hyperfekt** [comment]
Not just the GN flag, also the blacklist and the flags need to be changed in 73 - just as described in the commit you linked.

**Comment by amazari** [comment]
This new one add the missing patches from Arch relaxing libva version check while enforcing they are only applied on chromium 72.
It also actually set use_vaapi to true, an oversight in the first patch.

Sadly I must report that it still doesn't work as expected, h264 test videos are still decoded through software ffmpeg.
Can someone take a look at it ? Chromium takes ages to build on my current dev machine so someone with a better rig might have a quicker iterating cycle.
Thanks.

File: pkgs/applications/networking/browsers/chromium/common.nix [hunk 1]
```diff
@@ -127,6 +128,7 @@ let
       ++ optionals gnomeSupport [ gnome.GConf libgcrypt ]
       ++ optionals cupsSupport [ libgcrypt cups ]
       ++ optional pulseSupport libpulseaudio
+      ++ optional (VAAPISupport && (versionRange "72" "73")) libva
```
Comment (colemickens) [inline pkgs/applications/networking/browsers/chromium/common.nix]: Does versionRange clamp to 72-73? Don't we probably want it for all versions > 72 ?

**Comment by hyperfekt** [comment]
> Chromium takes ages to build on my current dev machine

This is also the reason I haven't taken this on yet, I've yet to set up remote builds on a c5.18xlarge instance. I will make this patch work if you haven't done so by then, but it is going to take me some time as I have another thing to finish first.

**Comment by colemickens** [comment]
I have a build that is built similarly, this is the error I get when I run: `GDK_BACKEND=x11 /nix/store/lxvj286d3iwh975ijczkcqnb9cbzn8gg-chromium-72.0.3626.121/bin/chromium --v=3 --log-level=0 --enable-logging=stderr &> foo.txt` and then search foo.txt for libva/vaapi:

```
[1077:1077:0305/090408.109028:VERBOSE1:va_stubs.cc(637)] dlopen(libva.so.2) failed, dlerror() says:
libva.so.2: cannot open shared object file: No such file or directory
```

`libva` is in `buildInputs`. Maybe Chromium is doing something in `va_stubs.cc` that we need to patch for use on Nix?

File: pkgs/applications/networking/browsers/chromium/common.nix (hunk 1 above)
Comment (amazari) [reply pkgs/applications/networking/browsers/chromium/common.nix]: yeah I guess we could replace that by a 
```suggestion
      ++ optional (VAAPISupport && (versionAtLeast version "73")) libva
```
as future versions will, whatever is the patchset required, need libva build dep.

**Comment by amazari** [comment]
@colemickens thanks for the input. I actually get the same log.

Looks like libva is linked at runtime through dlopen and not the usual linker. So the lib file aint in the standard LIBDIR and can't be resolved.
Reading from the manual, adding it to ''runtimeDependencies'' should make it visible at runtime through some LD_LIBRARY_PATH or linking magic.
I have a build going on overnight and will report back ASAP.

**Comment by colemickens** [comment]
@amazari I've done just that in my branch and I am progressing further (I'd had to do that already to enable Ozone). I have just been reluctant to post updates as I don't want to spam too much. Also, I fear that some other complications around libva (https://github.com/NixOS/nixpkgs/issues/55975) are related (I can cause similar errors as below with mpv+x11 but not with mpv+wayland.)

**Attempt 1** Naive:
```shell
[cole@xeep:~]$ unset LIBVA_DRIVER_NAME # it's not normally set
[cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
[cole@xeep:~]$ cat log.txt | grep libva
libva error: va_getDriverName() failed with unknown libva error,driver_name=(null)
```

**Attempt 1** Explicit: i965
```shell
[cole@xeep:~]$ ls /run/opengl-driver/lib/dri | grep drv_video | grep i965
i965_drv_video.la
i965_drv_video.so
[cole@xeep:~]$ export LIBVA_DRIVER_NAME=i965
[cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
[cole@xeep:~]$ cat log.txt | grep libva
libva error: /run/opengl-driver/lib/dri/i965_drv_video.so init failed
[17809:17809:0305/163001.156500:ERROR:vaapi_wrapper.cc(324)] vaInitialize failed: unknown libva error
```

I'm wondering if this is related to my local workaround for #55975.

**Attempt 2** Explicit: iHD
If you have `intel-media-driver` installed in nixos's opengl drivers:
```shell
[cole@xeep:~]$ ls /run/opengl-driver/lib/dri | grep drv_video | grep iHD
iHD_drv_video.so
[cole@xeep:~]$ export LIBVA_DRIVER_NAME=iHD
[cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
[cole@xeep:~]$ cat log.txt | grep libva
[cole@xeep:~]$ cat log.txt | grep --text vaapi
[19095:19095:0305/163315.855345:VERBOSE2:vaapi_video_encode_accelerator.cc(202)] VaapiVideoEncodeAccelerator(): 
[19095:19095:0305/163315.855375:VERBOSE2:vaapi_video_encode_accelerator.cc(206)] ~VaapiVideoEncodeAccelerator(): 
[19095:19095:0305/163315.855389:VERBOSE2:vaapi_jpeg_decode_accelerator.cc(306)] ~VaapiJpegDecodeAccelerator(): Destroying VaapiJpegDecodeAccelerator
[19095:19095:0305/163317.384730:VERBOSE2:vaapi_video_decode_accelerator.cc(198)] Initialize(): Initializing VAVDA, profile: h264 main
[19095:19312:0305/163317.385901:VERBOSE2:vaapi_video_decode_accelerator.cc(446)] DecodeTask(): Decoder requesting a new set of surfaces
[19095:19095:0305/163317.385941:VERBOSE2:vaapi_video_decode_accelerator.cc(499)] InitiateSurfaceSetChange(): Initiating surface set change
[19095:19095:0305/163317.385964:VERBOSE2:vaapi_video_decode_accelerator.cc(543)] TryFinishSurfaceSetChange(): Requesting 11 pictures of size: 1280x720
[19316:19316:0305/163317.489346:VERBOSE2:vaapi_video_encode_accelerator.cc(202)] VaapiVideoEncodeAccelerator(): 
[19316:19316:0305/163317.489375:VERBOSE2:vaapi_video_encode_accelerator.cc(206)] ~VaapiVideoEncodeAccelerator(): 
[19316:19316:0305/163317.489392:VERBOSE2:vaapi_jpeg_decode_accelerator.cc(306)] ~VaapiJpegDecodeAccelerator(): Destroying VaapiJpegDecodeAccelerator
```

[screenshot of media-internals with iHD](https://user-images.githubusercontent.com/327028/53847280-dedfd000-3f64-11e9-9f93-5eb275b5de96.png)


Does this mean it is working?

**Comment by amazari** [comment]
Hey @colemickens.
> **Attempt 1** Naive:
> 
> ```shell
> [cole@xeep:~]$ unset LIBVA_DRIVER_NAME # it's not normally set
> [cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
> [cole@xeep:~]$ cat log.txt | grep libva
> libva error: va_getDriverName() failed with unknown libva error,driver_name=(null)
> ```
> **Attempt 1** Explicit: i965
> 
> ```shell
> [cole@xeep:~]$ ls /run/opengl-driver/lib/dri | grep drv_video | grep i965
> i965_drv_video.la
> i965_drv_video.so
> [cole@xeep:~]$ export LIBVA_DRIVER_NAME=i965
> [cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
> [cole@xeep:~]$ cat log.txt | grep libva
> libva error: /run/opengl-driver/lib/dri/i965_drv_video.so init failed
> [17809:17809:0305/163001.156500:ERROR:vaapi_wrapper.cc(324)] vaInitialize failed: unknown libva error
> ```
> 
> I'm wondering if this is related to my local workaround for #55975.
> 

That's weird, at least libva is resolved and in the right library search path, which is better the current state of my tree.
I added 
`
runtimeDependencies = optional VAAPISupport [ libva ];
`
but the dlopen call is still failing. Would you mind sharing your solution?

> **Attempt 2** Explicit: iHD
> If you have `intel-media-driver` installed in nixos's opengl drivers:
> 
> ```shell
> [cole@xeep:~]$ ls /run/opengl-driver/lib/dri | grep drv_video | grep iHD
> iHD_drv_video.so
> [cole@xeep:~]$ export LIBVA_DRIVER_NAME=iHD
> [cole@xeep:~]$ GDK_BACKEND=x11 /nix/store/65n4m3rkn9q2bfcrmz2xpp61823v7cwm-chromium-72.0.3626.121/bin/chromium --log-level=9 --enable-logging=stderr --v=9 &> log.txt
> [cole@xeep:~]$ cat log.txt | grep libva
> [cole@xeep:~]$ cat log.txt | grep --text vaapi
> [19095:19095:0305/163315.855345:VERBOSE2:vaapi_video_encode_accelerator.cc(202)] VaapiVideoEncodeAccelerator(): 
> [19095:19095:0305/163315.855375:VERBOSE2:vaapi_video_encode_accelerator.cc(206)] ~VaapiVideoEncodeAccelerator(): 
> [19095:19095:0305/163315.855389:VERBOSE2:vaapi_jpeg_decode_accelerator.cc(306)] ~VaapiJpegDecodeAccelerator(): Destroying VaapiJpegDecodeAccelerator
> [19095:19095:0305/163317.384730:VERBOSE2:vaapi_video_decode_accelerator.cc(198)] Initialize(): Initializing VAVDA, profile: h264 main
> [19095:19312:0305/163317.385901:VERBOSE2:vaapi_video_decode_accelerator.cc(446)] DecodeTask(): Decoder requesting a new set of surfaces
> [19095:19095:0305/163317.385941:VERBOSE2:vaapi_video_decode_accelerator.cc(499)] InitiateSurfaceSetChange(): Initiating surface set change
> [19095:19095:0305/163317.385964:VERBOSE2:vaapi_video_decode_accelerator.cc(543)] TryFinishSurfaceSetChange(): Requesting 11 pictures of size: 1280x720
> [19316:19316:0305/163317.489346:VERBOSE2:vaapi_video_encode_accelerator.cc(202)] VaapiVideoEncodeAccelerator(): 
> [19316:19316:0305/163317.489375:VERBOSE2:vaapi_video_encode_accelerator.cc(206)] ~VaapiVideoEncodeAccelerator(): 
> [19316:19316:0305/163317.489392:VERBOSE2:vaapi_jpeg_decode_accelerator.cc(306)] ~VaapiJpegDecodeAccelerator(): Destroying VaapiJpegDecodeAccelerator
> ```
> 
> [screenshot of media-internals with iHD](https://user-images.githubusercontent.com/327028/53847280-dedfd000-3f64-11e9-9f93-5eb275b5de96.png)
> 
> Does this mean it is working?

It does look like libva intialization is successful but the screenshot contradicts that as it looks like FFmpegVideoDecoder is actually used. AFAIK, HW-accelerated decoding goes through MojoVideoDecoder or GpuVideoDecoder.
Did you patch the enable-mojo patch in ?

Thanks tor your time, dude

**Comment by amazari** [comment]
> That's weird, at least libva is resolved and in the right library search path, which is better the current state of my tree.
> I added
> `runtimeDependencies = optional VAAPISupport [ libva ];`
> but the dlopen call is still failing. Would you mind sharing your solution?

Ok, so I had a look at your patch and you dont use the `runtimeDependencies` mechanism which doesn't work in my case but an ad-hoc tweaking of the executable wrapper building. I am gonna try that.
It might be interesting to generalize and abstract that away in the long term to make it reusable.

**Comment by amazari** [comment]
Ok, so I obviously messed up with git. I am closing this PR and create a new one. Sorry about that.

**Pull request closed without merging**